src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Interop.Internal.Java.Util;
+using Microsoft.Spark.Interop;
+
+namespace Microsoft.Spark
+{
+    /// <summary>
+    /// A broadcast variable. Broadcast variables allow the programmer to keep a read-only variable
+    /// cached on each machine rather than shipping a copy of it with tasks. They can be used, for
+    /// example, to give every node a copy of a large input dataset in an efficient manner. Spark
+    /// also attempts to distribute broadcast variables using efficient broadcast algorithms to
+    /// reduce communication cost.
+    /// </summary>
+    [Serializable]
+    public sealed class Broadcast: IJvmObjectReferenceProvider
+    {
+        [NonSerialized]
+        private readonly JvmObjectReference _jvmObject;
+        private readonly long _bid;
+
+        internal Broadcast(SparkContext sc, object value, JvmObjectReference sparkContext)
+        {
+            var path = (string)Path.Combine(sc._temp_dir, Path.GetRandomFileName());
+            Version version = SparkEnvironment.SparkVersion;
+
+            // For Spark versions 2.3.0 and 2.3.1, Broadcast variable is created through different
+            // functions.
+            if (version.Major == 2 && version.Minor == 3 && (version.Build == 0 || 
```

Comment:
Is this big boolean used anywhere else?  I'm wondering if it should be some kind of utility function.

Reply:
So far it is not being used anywhere else. Do you think we should make it a utility function regardless?

Reply:
That's OK, it can stay here for now.